### ISSUE TITLE
kubelet fails to restart with CPUManager policy static when node's some CPU is offline

### ISSUE THREAD

**Issue by mtawaken** [kubernetes/kubernetes#124066]
### What happened?

I have a node with Configuration:
a. 96 CPUS with one CPU(ID=95) disabled. 
b. kubelet CPUManager static policy enabled.
c. kubeReserved: cpu: 2, systemReserved: cpu: 2

After start kubelet, the topology kubelet detected shows
`"Detected CPU topoloogy" topology=&{NumCPUs:95 NumCores:48 NumSockets:2 NumNUMANodes:2 CPUDetails:......`
Notes Core 25 contains CPU 25,73。 Core 47 contains CPU 47,95，but 95 is disabled and not shown in the log.

The static Policy process CPU reservation log below. 
`"takeFullCores: clainming core" core=47`
`"takeFullCores: clainming core" core=24`
`"takeRemainingCPUs: claiming CPU" cpu=25`
`"Reserved CPUs not available for exclusive assignment" reservedSize=4 reserved="24-25,47,72"`
`"Updated default CPUSet" cpuSet="0-94"`

The weird thing is when the first Pod admits, the reserved CPU(ID=25) is reused. 
`"Topology Affinity" pod="xxx" containerName="xxx" affinity={NUMANodeAffinity:<nil> Prefered:false}`
`"AllocateCPUs" numCPUs=20 socket=<nil>`
`"takeFullCores: clainming core" core=25`
`"takeRemainingCPUs: claiming CPU" cpu=26`
`"takeRemainingCPUs: claiming CPU" cpu=74`
....
`"Updated default CPUSet" cpuSet="0-24,35-72,83-94"`
`"AllocateCPUs" result="25-34,73-82"`

The confliction would further cause kubelet restart failure because of the static policy validation
`"Static policy invalid state, please drain node and remove policy state file" err="not all reserved cpus: \"24-25,47,72\" are present in defaultCpuSet: \"0-24,35-72,83-94\""`

### What did you expect to happen?

Pod Cpuset Allocation should not share with Reserved CPUs

### How can we reproduce it (as minimally and precisely as possible)?

1. use `chcpu -d` to disable one CPU in the node. Check `/sys/devices/system/cpu/online` file to make sure it works.
2. change kubelet configuration
   a. cpuManagerPolicy: static
   b. kubeReserved: cpu: 2, systemReserved: cpu: 2
3. restart kubelet
4. create a guaranteed Pod with some CPU
5. restart kubelet, expects it would fail to startup

### Anything else we need to know?

it seems that kubelet CPUManager static Policy have not taken CPU offline senario into account. 
1. in the specific case above, this method would result in 1(95/48), not 2
https://github.com/kubernetes/kubernetes/blob/227c2e7c2b2c05a9c8b2885460e28e4da25cf558/pkg/kubelet/cm/cpumanager/topology/topology.go#L49-L54
2. thus result in a Used Core as free (1==1)
https://github.com/kubernetes/kubernetes/blob/227c2e7c2b2c05a9c8b2885460e28e4da25cf558/pkg/kubelet/cm/cpumanager/cpu_assignment.go#L259-L262
3. all logic releated with CPUsPerCore, CPUsPerSocket is affected

### Kubernetes version

<details>
tested in kubernetes: v1.25, other version is the same.
</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by AnishShah** [comment]
/kind feature
/remove-kind bug

kubelet does not respond well to dynamic changes to resources on the node. There is an open KEP to improve this with dynamic changes to resources.

**Comment by haircommander** [comment]
a bit more context on @AnishShah 's comment:

In the SIG-Node CI meeting today we talked about this and came to the conclusion that this is a feature request. In part because of the structured error that is clearly catching this problem, and in part because this is ultimately a hotplugged resource problem, even though the kubelet is restarting in between. How the kubelet responds to cpus being taken offline underneath it would be solved the same way if the kubelet was continuously running or if it was restarted before and after. 

I thought there was an open KEP about hotplugging memory/cpu and having the kubelet respond, but I can't find it.

We also talked a bit about what the kubelet should do in this case. We decided the kubelet should probably evict in this case, but that means making the eviction manager aware of the needs of the cpu manager. 

All of this together informed the decision to mark this as a feature

**Comment by mtawaken** [comment]
thanks for the comment. 
I went through the webinar recording and agree that the hotplugged resource feature requires a lot of effort and no need to be in haste. 
BUT I think there's some aspect missing in the discussion. In this case, the node configuration is static, no dynamic change involved. kubelet is wrong from the very beginning(pls refer the code I pasted above). 
So maybe it's still a bug unless we declare not considering odd number of CPUs with HT?

PS: the node's one CPU is always offline because of the HPC issue, it is necessary for tunning in some AMD architectures.

**Comment by haircommander** [comment]
I think "very beginning" is questionable, since you have running containers and an existing state file. I get the point that it's the beginning of the kubelet's process execution, but clearly this node has not just been rebooted or started from scratch. That's why it feels like a hotplug case to me.

**Comment by ffromani** [comment]
I think this is what @haircommander mentioned before: https://github.com/kubernetes/enhancements/issues/3953

**Comment by mtawaken** [comment]
> I think "very beginning" is questionable, since you have running containers and an existing state file
Understood the case that dynamic resource changes and then restart kubelet. About this case, we could reproduce it with no running containers and clear any existing state file before starting kubelet, it is technically a brand new node.

To clarify,
* Dynaminc node resize Feature: the Topology is not up-to-date, and we need to inform Topology and make it reasonable about existing allocations and further allocations
* This issue: the Topology is already up-to-date, no offline CPU is used, but further container allocation shares some CPU with reserved CPUs(kube-reserved/system-reserved). It mainly violates the primary design.

**Comment by ffromani** [comment]
> @ffromani sorry for the late reply. I evaluated `full-pcpus-only` option, it only works as an extend option for "distribute-cpus-across-numa", and still relys on the the result of `CPUsPerCore`, which is already wrong.

Right. The critical flow is that we do trivial divisions in the cpu topology code (and in the surrounding code) rather than enumerating the CPUs and taking their relationship (and online/offline status) intou account.

> I submit a pr as the workaround inside out environment by modifying IsCoreFree/IsSocketFree method. It is not a full solution for the offline CPU scenario(maybe should covered by the dynamic node resize feature?), but it can solve the reuse problem of reserved CPUs, and make kubelet restarting back to normal.

This seems promising, I'll have a look

**Comment by ffromani** [comment]
/kind bug
/priority important-longterm

I believe this is a legit bug, albeit I expect to be pretty rare in practice. Retriaging as such.

**Comment by sftim** [comment]
Similar to https://github.com/kubernetes/kubernetes/issues/71622 BTW

**Comment by com30n** [comment]
> I expect to be pretty rare in practice

We've faced the error on several hosts,
which is quite strange because all our nodes use cpumanager, and we have to have downtime to replace one of the cpu, even though it works fine from a linux perspective (382 cores out of 384 are working correctly), but kubelet is not working.

**Comment by ffromani** [comment]
> > I expect to be pretty rare in practice
> 
> We've faced the error on several hosts, which is quite strange because all our nodes use cpumanager, and we have to have downtime to replace one of the cpu, even though it works fine from a linux perspective (382 cores out of 384 are working correctly), but kubelet is not working.

ok, good to know. point taken. Thanks for reporting!
